### ISSUE TITLE
CCG gets stuck

### ISSUE THREAD

**Issue by beyeler** [marketdesignresearch/mechlib#11]
CCG payment computation in MRVM get sometimes stuck. I saw that in the old sats code there was an exit condition for this loop that it exits when the allocation value does not increase due to some rounding errors (CPLEX?). May we need here to add something similar. However, I'm not yet a CCG expert so I would appreciate when you could have a look at it.

https://github.com/marketdesignresearch/mechlib/blob/97713d744fa0def407c137cfd7f2b2bdd9cbaff1/src/main/java/org/marketdesignresearch/mechlib/outcomerules/ccg/CCGOutcomeRule.java#L87

**Comment by islerfab** [comment]
@beyeler Do you know if the latest changes regarding scaling already solved this? Just to know. I'd still like to solve that in a cleaner way when I get to it.

**Comment by beyeler** [comment]
No the issue seems not to be related to be related to the scaling changes. I expect the issue to be related to numerical isses that arrise from many consecutive CPLEX runs. I debuged one version that run in an infinite loop here and discovered the following values:

Value of most Blocking Allocation: 403129812.3409271997....
Value of winner payments:            403129812.3408643613.....

The precision of the comparison is 0.000010. In sats this issue was resolved by stopping the algorithm when the most blocking allocation value does not advance anymore. However I decided to implement a slighly less restrictive implementation. As long as the value of the most blocking allocation does not advance the precision of the comparision is reduced by a factor of 10. Therefore we still have a guarantee that payments are very close to the current most blocking coalition value.
